pkg/server/bundle/pubmanager/pubmanager.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/spiffe/spire/proto/spire/common"
+)
+
+const (
+	// refreshInterval is the interval to check for an updated trust bundle.
+	refreshInterval = 30 * time.Second
+)
+
+// NewManager creates a new bundle publishing manager.
+func NewManager(c *ManagerConfig) *Manager {
+	return newManager(c)
+}
+
+// ManagerConfig is the config for the bundle publishing manager.
+type ManagerConfig struct {
+	BundlePublishers []bundlepublisher.BundlePublisher
+	DataStore        datastore.DataStore
+	Clock            clock.Clock
+	Log              logrus.FieldLogger
+	TrustDomain      spiffeid.TrustDomain
+}
+
+// Manager is the manager for bundle publishing. It implements the PubManager
+// interface.
+type Manager struct {
+	bundleUpdatedCh  chan struct{}
+	bundlePublishers []bundlepublisher.BundlePublisher
+	clock            clock.Clock
+	dataStore        datastore.DataStore
+	log              logrus.FieldLogger
+	trustDomain      spiffeid.TrustDomain
+
+	hooks struct {
+		// Test hook used to indicate an attempt to publish a bundle using a
+		// specific bundle publisher.
+		publishResultCh chan *publishResult
+
+		// Test hook used to indicate when the action of publishing a bundle
+		// has finished.
+		publishedCh chan error
+	}
+}
+
+// Run runs the bundle publishing manager.
+func (m *Manager) Run(ctx context.Context) error {
+	if m.dataStore == nil {
+		return errors.New("missing datastore")
+	}
+
+	if m.trustDomain.IsZero() {
+		return errors.New("missing trust domain")
+	}
+
+	ticker := m.clock.Ticker(refreshInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			m.callPublishBundle(ctx)
+		case <-m.bundleUpdatedCh:
+			m.callPublishBundle(ctx)
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// BundleUpdated tells the bundle publishing manager that the bundle has been
+// updated and forces a PublishBundle operation on all the plugins.
+func (m *Manager) BundleUpdated() {
+	m.drainBundleUpdated()
+
+	if m.bundleUpdatedCh != nil {
+		m.bundleUpdatedCh <- struct{}{}
+	}
+}
+
+// callPublishBundle calls the publishBundle function and logs if there was an
+// error.
+func (m *Manager) callPublishBundle(ctx context.Context) {
+	if err := m.publishBundle(ctx); err != nil && ctx.Err() == nil {
+		m.log.WithError(err).Error("Failed to publish bundle")
+	}
+}
+
+// publishBundle iterates through the configured bundle publishers and calls
+// PublishBundle with the fetched bundle. This function only returns an error
+// if bundle publishers can't be called due to a failure fetching the bundle
+// from the datastore.
+func (m *Manager) publishBundle(ctx context.Context) (err error) {
+	defer func() {
+		m.publishDone(err)
+	}()
+
+	if len(m.bundlePublishers) == 0 {
+		return nil
+	}
+
+	bundle, err := m.dataStore.FetchBundle(ctx, m.trustDomain.IDString())
+	if err != nil {
+		return fmt.Errorf("failed to fetch bundle from datastore: %w", err)
+	}
+
+	for _, bp := range m.bundlePublishers {
+		bp := bp
+		go func() {
```

Comment:
Should we be waiting for these goroutines to exit before allowing the function to return?

Reply:
Good point, I think it should.